meta-balena-common/recipes-core/initrdscripts/files/cryptsetup

```diff
@@ -18,12 +26,41 @@ cryptsetup_enabled() {
 }
 
 cryptsetup_run() {
+    # Die if anything fails here
+    set -e
+
     EFI_MOUNT_DIR="/efi"
     mkdir "$EFI_MOUNT_DIR"
-    mount /dev/disk/by-partlabel/balena-efi "$EFI_MOUNT_DIR"
+    mount "$EFI_DEV" "$EFI_MOUNT_DIR"
 
     PASSPHRASE_FILE=/balena-luks.pwd
-    tpm2_rsadecrypt -c "$EFI_MOUNT_DIR/balena-luks.ctx" -o "$PASSPHRASE_FILE" "$EFI_MOUNT_DIR/balena-luks.enc"
+    if hw_decrypt_passphrase "$EFI_MOUNT_DIR" "0,1,2,3" "$PASSPHRASE_FILE"; then
+        info "Successfully unlocked LUKS passphrase using the TPM"
+    elif hw_decrypt_passphrase "$EFI_MOUNT_DIR" "0,2,3" "$PASSPHRASE_FILE"; then
+        info "Unlocked LUKS passphrase without UEFI config protection, re-locking"
+
+        RESULT_DIR="/balena-luks"
+        mkdir -p "$RESULT_DIR"
+
+        # Create a PCR protection policy
+        # We are using PCRs 0, 1, 2 and 3
+        # This ensures secure boot is enabled and no UEFI configuration has been tampered with
+        hw_encrypt_passphrase "$PASSPHRASE_FILE" "0,1,2,3" "$RESULT_DIR"
+
+        # Ditch the old key pair from the TPM and replace existing files with new ones
+        tpm2_evictcontrol -c "$EFI_MOUNT_DIR/balena-luks.ctx"
+        mv "$RESULT_DIR/persistent.ctx" "$EFI_MOUNT_DIR/balena-luks.ctx"
+        mv "$RESULT_DIR/passphrase.enc" "$EFI_MOUNT_DIR/balena-luks.enc"
```

Comment:
nitpick, maybe sync each operation to minimise the chances of losing data on power loss:
```suggestion
        mv "$RESULT_DIR/persistent.ctx" "$EFI_MOUNT_DIR/balena-luks.ctx" && sync
        mv "$RESULT_DIR/passphrase.enc" "$EFI_MOUNT_DIR/balena-luks.enc" && sync
```

Reply:
Added